src/libopensc/card-piv.c

```diff
@@ -2440,15 +2440,16 @@ static int piv_select_file(sc_card_t *card, const sc_path_t *in_path,
 	pathlen = in_path->len;
 
 	/* only support single EF in current application */
+	/* 
+	 * PIV emulates files, and only does so becauses sc_pkcs15_* uses
+	 * select_file and read_binary. The emulation adds path emulated structures
+	 * so piv_select_file will find it.
+	 * there is no dir. Only direct access to emulated files
+	 * thus opensc-tool and opensc-explorer can not read the emulated files
+	 */
 
 	if (memcmp(path, "\x3F\x00", 2) == 0) {
-		if (pathlen == 2)   {
-			r = piv_select_aid(card, piv_aids[0].value, piv_aids[0].len_short, NULL, NULL);
-			SC_TEST_RET(card->ctx, SC_LOG_DEBUG_NORMAL, r, "Cannot select PIV AID");
-
-			SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_NORMAL, r);
-		}
-		else if (pathlen > 2) {
+		if (pathlen > 2) {
```

Comment:
Previous reply said SC_TET_RET, its the SC_FUNC_RETURN  that returns 0 and without setting *file

On 10/16/2015 4:44 AM, Frank Morgner wrote:
> In src/libopensc/card-piv.c <https://github.com/OpenSC/OpenSC/pull/579#discussion_r42225468>:
>
>>
>>  	if (memcmp(path, "\x3F\x00", 2) == 0) {
>> -		if (pathlen == 2)   {
>> -			r = piv_select_aid(card, piv_aids[0].value, piv_aids[0].len_short, NULL, NULL);
>> -			SC_TEST_RET(card->ctx, SC_LOG_DEBUG_NORMAL, r, "Cannot select PIV AID");
>> -
>> -			SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_NORMAL, r);
>> -		}
>> -		else if (pathlen > 2) {
>> +		if (pathlen > 2) {
>
> The code you remove /should/ work. What was the problem?
>
> The change you introduced will /always/ remove the first FID. This will not work with a path relative to the current DF.
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/pull/579/files#r42225468>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>